scheduler/feasible.go

```diff
@@ -192,14 +204,29 @@ func (iter *ProposedAllocConstraintIterator) Next() *structs.Node {
 		// Get the next option from the source
 		option := iter.source.Next()
 
-		// Hot-path if the option is nil or there are no distinct_hosts constraints.
-		if option == nil || !(iter.jobDistinctHosts || iter.tgDistinctHosts) {
+		// Hot-path if the option is nil or there are no distinct_hosts or
+		// distinct_property constraints.
+		hosts := iter.jobDistinctHosts || iter.tgDistinctHosts
+		properties := iter.distinctProperties.HasDistinctPropertyConstraints()
+		if option == nil || !(hosts || properties) {
 			return option
 		}
 
-		if !iter.satisfiesDistinctHosts(option) {
-			iter.ctx.Metrics().FilterNode(option, structs.ConstraintDistinctHosts)
-			continue
+		// Check if the host constraints are satisfied
+		if hosts {
+			if !iter.satisfiesDistinctHosts(option) {
+				iter.ctx.Metrics().FilterNode(option, structs.ConstraintDistinctHosts)
+				continue
+			}
+		}
+
+		// Check if the property constraints are satisfied
+		if properties {
```

Comment:
I wonder if this should just be a new iterator? Seem's odd to create a split code path like this, instead of just having another iterator type.